File: android/app/src/main/java/com/kvkleh/sbtsurvey/ui/components/FormComponents.kt
package com.kvkleh.sbtsurvey.ui.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.selection.selectable
import androidx.compose.foundation.selection.selectableGroup
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowDropDown
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.OutlinedTextFieldDefaults
import androidx.compose.material3.RadioButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp

/**
 * A numbered section of the survey form. Each section is its own card so that the
 * form reads as a checklist while scrolling.
 */
@Composable
fun SectionCard(
    number: Int?,
    title: String,
    modifier: Modifier = Modifier,
    subtitle: String? = null,
    icon: ImageVector? = null,
    trailing: (@Composable () -> Unit)? = null,
    content: @Composable ColumnScope.() -> Unit
) {
    Card(
        modifier = modifier.fillMaxWidth(),
        shape = MaterialTheme.shapes.medium,
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.surfaceContainerLowest
        ),
        elevation = CardDefaults.cardElevation(defaultElevation = 1.dp)
    ) {
        Column(modifier = Modifier.padding(horizontal = 16.dp, vertical = 16.dp)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                if (number != null) {
                    Box(
                        modifier = Modifier
                            .size(30.dp)
                            .background(MaterialTheme.colorScheme.primaryContainer, CircleShape),
                        contentAlignment = Alignment.Center
                    ) {
                        Text(
                            text = number.toString(),
                            style = MaterialTheme.typography.labelLarge,
                            color = MaterialTheme.colorScheme.onPrimaryContainer
                        )
                    }
                    Spacer(Modifier.width(12.dp))
                } else if (icon != null) {
                    Icon(
                        imageVector = icon,
                        contentDescription = null,
                        tint = MaterialTheme.colorScheme.primary
                    )
                    Spacer(Modifier.width(12.dp))
                }
                Column(modifier = Modifier.weight(1f)) {
                    Text(
                        text = title,
                        style = MaterialTheme.typography.titleMedium,
                        color = MaterialTheme.colorScheme.onSurface
                    )
                    if (subtitle != null) {
                        Text(
                            text = subtitle,
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                    }
                }
                if (trailing != null) trailing()
            }
            Spacer(Modifier.height(14.dp))
            Column(verticalArrangement = Arrangement.spacedBy(12.dp), content = content)
        }
    }
}

/** Outlined text field with a required marker, helper text and a large touch target. */
@Composable
fun SbtTextField(
    value: String,
    onValueChange: (String) -> Unit,
    label: String,
    modifier: Modifier = Modifier,
    required: Boolean = false,
    supportingText: String? = null,
    errorText: String? = null,
    keyboardType: KeyboardType = KeyboardType.Text,
    imeAction: ImeAction = ImeAction.Next,
    singleLine: Boolean = true,
    enabled: Boolean = true,
    trailingIcon: (@Composable () -> Unit)? = null
) {
    Column(modifier = modifier.fillMaxWidth()) {
        OutlinedTextField(
            value = value,
            onValueChange = onValueChange,
            modifier = Modifier.fillMaxWidth(),
            label = { Text(if (required) "$label *" else label) },
            singleLine = singleLine,
            enabled = enabled,
            isError = errorText != null,
            shape = MaterialTheme.shapes.small,
            trailingIcon = trailingIcon,
            keyboardOptions = KeyboardOptions(keyboardType = keyboardType, imeAction = imeAction),
            colors = OutlinedTextFieldDefaults.colors(
                focusedContainerColor = MaterialTheme.colorScheme.surfaceContainerLowest,
                unfocusedContainerColor = MaterialTheme.colorScheme.surfaceContainerLowest,
                disabledContainerColor = MaterialTheme.colorScheme.surfaceContainerLowest,
                disabledTextColor = MaterialTheme.colorScheme.onSurface,
                disabledLabelColor = MaterialTheme.colorScheme.onSurfaceVariant,
                disabledBorderColor = MaterialTheme.colorScheme.outline
            ),
            textStyle = MaterialTheme.typography.bodyLarge
        )
        val helper = errorText ?: supportingText
        if (helper != null) {
            Text(
                text = helper,
                style = MaterialTheme.typography.bodySmall,
                color = if (errorText != null) {
                    MaterialTheme.colorScheme.error
                } else {
                    MaterialTheme.colorScheme.onSurfaceVariant
                },
                modifier = Modifier.padding(start = 14.dp, top = 4.dp)
            )
        }
    }
}

/**
 * A text field with a suggestion list. Values can be picked from the list or typed
 * by hand, which is what field work needs: a village that is not on any list must
 * still be recordable.
 */
@Composable
fun SbtDropdownField(
    value: String,
    onValueChange: (String) -> Unit,
    label: String,
    options: List<String>,
    modifier: Modifier = Modifier,
    required: Boolean = false,
    supportingText: String? = null,
    errorText: String? = null,
    allowManualEntry: Boolean = true
) {
    var expanded by remember { mutableStateOf(false) }
    Box(modifier = modifier.fillMaxWidth()) {
        SbtTextField(
            value = value,
            onValueChange = { if (allowManualEntry) onValueChange(it) },
            label = label,
            required = required,
            supportingText = supportingText,
            errorText = errorText,
            enabled = allowManualEntry,
            trailingIcon = {
                IconButton(onClick = { expanded = true }) {
                    Icon(
                        imageVector = Icons.Filled.ArrowDropDown,
                        contentDescription = "Show $label options"
                    )
                }
            }
        )
        if (!allowManualEntry) {
            // A read-only field still has to open its list when tapped anywhere.
            Box(
                modifier = Modifier
                    .matchParentSize()
                    .clickable { expanded = true }
            )
        }
        DropdownMenu(
            expanded = expanded,
            onDismissRequest = { expanded = false },
            modifier = Modifier
                .heightIn(max = 340.dp)
                .background(MaterialTheme.colorScheme.surfaceContainerLowest)
        ) {
            if (options.isEmpty()) {
                DropdownMenuItem(
                    text = { Text("No suggestions – type the name") },
                    onClick = { expanded = false },
                    enabled = false
                )
            }
            options.forEach { option ->
                DropdownMenuItem(
                    text = {
                        Text(
                            text = option,
                            style = MaterialTheme.typography.bodyLarge,
                            fontWeight = if (option == value) FontWeight.SemiBold else FontWeight.Normal
                        )
                    },
                    onClick = {
                        onValueChange(option)
                        expanded = false
                    }
                )
            }
        }
    }
}

/**
 * Single-choice list rendered as full-width rows. Radio rows beat a compact
 * dropdown outdoors: the whole row is a 56 dp target that can be hit with gloves.
 */
@Composable
fun SbtRadioGroup(
    options: List<String>,
    selected: String?,
    onSelected: (String) -> Unit,
    modifier: Modifier = Modifier,
    errorText: String? = null
) {
    Column(modifier = modifier.selectableGroup()) {
        options.forEach { option ->
            val isSelected = option == selected
            Surface(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(vertical = 3.dp)
                    .heightIn(min = 56.dp)
                    .selectable(
                        selected = isSelected,
                        onClick = { onSelected(option) },
                        role = Role.RadioButton
                    ),
                shape = MaterialTheme.shapes.small,
                color = if (isSelected) {
                    MaterialTheme.colorScheme.primaryContainer
                } else {
                    MaterialTheme.colorScheme.surfaceContainerLow
                },
                border = BorderStroke(
                    width = if (isSelected) 1.5.dp else 1.dp,
                    color = if (isSelected) {
                        MaterialTheme.colorScheme.primary
                    } else {
                        MaterialTheme.colorScheme.outlineVariant
                    }
                )
            ) {
                Row(
                    modifier = Modifier.padding(horizontal = 12.dp, vertical = 8.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    RadioButton(selected = isSelected, onClick = null)
                    Spacer(Modifier.width(10.dp))
                    Text(
                        text = option,
                        style = MaterialTheme.typography.bodyLarge,
                        fontWeight = if (isSelected) FontWeight.SemiBold else FontWeight.Normal,
                        color = if (isSelected) {
                            MaterialTheme.colorScheme.onPrimaryContainer
                        } else {
                            MaterialTheme.colorScheme.onSurface
                        }
                    )
                }
            }
        }
        if (errorText != null) {
            Text(
                text = errorText,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.error,
                modifier = Modifier.padding(start = 4.dp, top = 6.dp)
            )
        }
    }
}

/** Compact segmented control, used for the plant height unit. */
@Composable
fun SbtSegmentedToggle(
    options: List<String>,
    labels: List<String>,
    selected: String,
    onSelected: (String) -> Unit,
    modifier: Modifier = Modifier
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .height(52.dp)
            .border(
                width = 1.dp,
                color = MaterialTheme.colorScheme.outlineVariant,
                shape = MaterialTheme.shapes.small
            )
            .padding(4.dp),
        horizontalArrangement = Arrangement.spacedBy(4.dp)
    ) {
        options.forEachIndexed { index, option ->
            val isSelected = option == selected
            Surface(
                modifier = Modifier
                    .weight(1f)
                    .fillMaxHeight()
                    .selectable(
                        selected = isSelected,
                        onClick = { onSelected(option) },
                        role = Role.RadioButton
                    ),
                shape = RoundedCornerShape(9.dp),
                color = if (isSelected) {
                    MaterialTheme.colorScheme.primary
                } else {
                    MaterialTheme.colorScheme.surfaceContainerLow
                }
            ) {
                Box(contentAlignment = Alignment.Center) {
                    Text(
                        text = labels.getOrElse(index) { option },
                        style = MaterialTheme.typography.labelLarge,
                        textAlign = TextAlign.Center,
                        color = if (isSelected) {
                            MaterialTheme.colorScheme.onPrimary
                        } else {
                            MaterialTheme.colorScheme.onSurfaceVariant
                        }
                    )
                }
            }
        }
    }
}

/** A labelled key/value line, used on the review and detail screens. */
@Composable
fun DetailRow(
    label: String,
    value: String,
    modifier: Modifier = Modifier,
    emphasise: Boolean = false
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .padding(vertical = 6.dp),
        verticalAlignment = Alignment.Top
    ) {
        Text(
            text = label,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.weight(0.45f)
        )
        Text(
            text = value.ifBlank { "—" },
            style = MaterialTheme.typography.bodyLarge,
            fontWeight = if (emphasise) FontWeight.SemiBold else FontWeight.Normal,
            color = if (value.isBlank()) {
                MaterialTheme.colorScheme.onSurfaceVariant
            } else {
                MaterialTheme.colorScheme.onSurface
            },
            modifier = Modifier.weight(0.55f)
        )
    }
}

/** Scrolling container shared by the form-style screens. */
@Composable
fun FormColumn(
    modifier: Modifier = Modifier,
    contentPadding: PaddingValues = PaddingValues(16.dp),
    content: @Composable ColumnScope.() -> Unit
) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .verticalScroll(rememberScrollState())
            .padding(contentPadding),
        verticalArrangement = Arrangement.spacedBy(14.dp),
        content = content
    )
}

File: ios/SeabuckthornSurvey/UI/Components/FormComponents.swift
import SwiftUI

/// A numbered section of the survey form. Each section is its own card so the form
/// reads as a checklist while scrolling.
struct SectionCard<Content: View>: View {

    var number: Int?
    var symbol: String?
    var title: String
    var subtitle: String?
    var trailing: AnyView?
    @ViewBuilder var content: () -> Content

    init(
        number: Int? = nil,
        symbol: String? = nil,
        title: String,
        subtitle: String? = nil,
        trailing: AnyView? = nil,
        @ViewBuilder content: @escaping () -> Content
    ) {
        self.number = number
        self.symbol = symbol
        self.title = title
        self.subtitle = subtitle
        self.trailing = trailing
        self.content = content
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 14) {
            HStack(alignment: .center, spacing: 12) {
                if let number {
                    Text("\(number)")
                        .font(.subheadline.weight(.semibold))
                        .foregroundStyle(SBTColor.onPrimaryContainer)
                        .frame(width: 30, height: 30)
                        .background(SBTColor.primaryContainer, in: Circle())
                } else if let symbol {
                    Image(systemName: symbol)
                        .font(.title3)
                        .foregroundStyle(SBTColor.primary)
                }
                VStack(alignment: .leading, spacing: 2) {
                    Text(title)
                        .font(.headline)
                        .foregroundStyle(SBTColor.ink)
                    if let subtitle {
                        Text(subtitle)
                            .font(.caption)
                            .foregroundStyle(SBTColor.inkSoft)
                    }
                }
                Spacer(minLength: 8)
                if let trailing { trailing }
            }
            VStack(alignment: .leading, spacing: 12) {
                content()
            }
        }
        .padding(16)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(SBTColor.surface, in: RoundedRectangle(cornerRadius: SBTMetrics.cornerRadius))
        .overlay(
            RoundedRectangle(cornerRadius: SBTMetrics.cornerRadius)
                .stroke(SBTColor.outline.opacity(0.6), lineWidth: 1)
        )
    }
}

/// Text field with a required marker, helper text and a large touch target.
struct SBTTextField: View {

    let title: String
    @Binding var text: String
    var required: Bool = false
    var supportingText: String?
    var errorText: String?
    var keyboard: UIKeyboardType = .default
    var autocapitalization: TextInputAutocapitalization = .words

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(required ? "\(title) *" : title)
                .font(.subheadline.weight(.medium))
                .foregroundStyle(SBTColor.inkSoft)
            TextField("", text: $text)
                .font(.body)
                .keyboardType(keyboard)
                .textInputAutocapitalization(autocapitalization)
                .autocorrectionDisabled()
                .padding(.horizontal, 14)
                .frame(minHeight: SBTMetrics.touchTarget)
                .background(
                    SBTColor.surface,
                    in: RoundedRectangle(cornerRadius: SBTMetrics.fieldCornerRadius)
                )
                .overlay(
                    RoundedRectangle(cornerRadius: SBTMetrics.fieldCornerRadius)
                        .stroke(errorText == nil ? SBTColor.outline : SBTColor.danger, lineWidth: 1.5)
                )
            if let helper = errorText ?? supportingText {
                Text(helper)
                    .font(.caption)
                    .foregroundStyle(errorText == nil ? SBTColor.inkSoft : SBTColor.danger)
            }
        }
    }
}

/// A text field with a suggestion menu. Values can be picked from the list or typed
/// by hand, which is what field work needs: a village that is on no list must still
/// be recordable.
struct SBTPickerField: View {

    let title: String
    @Binding var text: String
    let options: [String]
    var required: Bool = false
    var supportingText: String?
    var errorText: String?

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(required ? "\(title) *" : title)
                .font(.subheadline.weight(.medium))
                .foregroundStyle(SBTColor.inkSoft)
            HStack(spacing: 0) {
                TextField("", text: $text)
                    .font(.body)
                    .textInputAutocapitalization(.words)
                    .autocorrectionDisabled()
                    .padding(.leading, 14)
                Menu {
                    if options.isEmpty {
                        Text("No suggestions – type the name")
                    }
                    ForEach(options, id: \.self) { option in
                        Button {
                            text = option
                        } label: {
                            if option == text {
                                Label(option, systemImage: "checkmark")
                            } else {
                                Text(option)
                            }
                        }
                    }
                } label: {
                    Image(systemName: "chevron.down")
                        .font(.body.weight(.semibold))
                        .foregroundStyle(SBTColor.primary)
                        .frame(width: 52, height: SBTMetrics.touchTarget)
                        .contentShape(Rectangle())
                }
                .accessibilityLabel("Show \(title) options")
            }
            .frame(minHeight: SBTMetrics.touchTarget)
            .background(
                SBTColor.surface,
                in: RoundedRectangle(cornerRadius: SBTMetrics.fieldCornerRadius)
            )
            .overlay(
                RoundedRectangle(cornerRadius: SBTMetrics.fieldCornerRadius)
                    .stroke(errorText == nil ? SBTColor.outline : SBTColor.danger, lineWidth: 1.5)
            )
            if let helper = errorText ?? supportingText {
                Text(helper)
                    .font(.caption)
                    .foregroundStyle(errorText == nil ? SBTColor.inkSoft : SBTColor.danger)
            }
        }
    }
}

/// Single-choice list rendered as full-width rows. Rows beat a compact picker
/// outdoors: the whole row is a large target that can be hit with gloves.
struct SBTChoiceRows: View {

    let options: [String]
    @Binding var selection: String?
    var errorText: String?

    var body: some View {
        VStack(spacing: 6) {
            ForEach(options, id: \.self) { option in
                let isSelected = option == selection
                Button {
                    selection = option
                } label: {
                    HStack(spacing: 12) {
                        Image(systemName: isSelected ? "largecircle.fill.circle" : "circle")
                            .font(.title3)
                            .foregroundStyle(isSelected ? SBTColor.primary : SBTColor.inkSoft)
                        Text(option)
                            .font(.body.weight(isSelected ? .semibold : .regular))
                            .foregroundStyle(isSelected ? SBTColor.onPrimaryContainer : SBTColor.ink)
                        Spacer()
                    }
                    .padding(.horizontal, 14)
                    .frame(minHeight: SBTMetrics.touchTarget)
                    .background(
                        isSelected ? SBTColor.primaryContainer : SBTColor.surface,
                        in: RoundedRectangle(cornerRadius: SBTMetrics.fieldCornerRadius)
                    )
                    .overlay(
                        RoundedRectangle(cornerRadius: SBTMetrics.fieldCornerRadius)
                            .stroke(
                                isSelected ? SBTColor.primary : SBTColor.outline,
                                lineWidth: isSelected ? 1.5 : 1
                            )
                    )
                }
                .buttonStyle(.plain)
                .accessibilityAddTraits(isSelected ? [.isSelected] : [])
            }
            if let errorText {
                Text(errorText)
                    .font(.caption)
                    .foregroundStyle(SBTColor.danger)
                    .frame(maxWidth: .infinity, alignment: .leading)
            }
        }
    }
}

/// Compact segmented control, used for the plant height unit.
struct SBTSegmentedToggle: View {

    let options: [String]
    let labels: [String]
    @Binding var selection: String

    var body: some View {
        HStack(spacing: 4) {
            ForEach(Array(options.enumerated()), id: \.element) { index, option in
                let isSelected = option == selection
                Button {
                    selection = option
                } label: {
                    Text(labels.indices.contains(index) ? labels[index] : option)
                        .font(.subheadline.weight(.semibold))
                        .foregroundStyle(isSelected ? Color.white : SBTColor.inkSoft)
                        .frame(maxWidth: .infinity)
                        .frame(height: 44)
                        .background(
                            isSelected ? SBTColor.primary : SBTColor.surface,
                            in: RoundedRectangle(cornerRadius: 9)
                        )
                }
                .buttonStyle(.plain)
            }
        }
        .padding(4)
        .overlay(
            RoundedRectangle(cornerRadius: SBTMetrics.fieldCornerRadius)
                .stroke(SBTColor.outline, lineWidth: 1)
        )
    }
}

/// A labelled key/value line, used on the review and detail screens.
struct DetailRow: View {

    let label: String
    let value: String
    var emphasise: Bool = false

    init(_ label: String, _ value: String, emphasise: Bool = false) {
        self.label = label
        self.value = value
        self.emphasise = emphasise
    }

    var body: some View {
        HStack(alignment: .top, spacing: 12) {
            Text(label)
                .font(.subheadline)
                .foregroundStyle(SBTColor.inkSoft)
                .frame(maxWidth: .infinity, alignment: .leading)
            Text(value.isEmpty ? "—" : value)
                .font(.body.weight(emphasise ? .semibold : .regular))
                .foregroundStyle(value.isEmpty ? SBTColor.inkSoft : SBTColor.ink)
                .frame(maxWidth: .infinity, alignment: .leading)
        }
        .padding(.vertical, 3)
    }
}

/// The primary action button used at the bottom of each step.
struct PrimaryButton: View {

    let title: String
    var systemImage: String?
    var isBusy: Bool = false
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            HStack(spacing: 10) {
                if isBusy {
                    ProgressView().tint(.white)
                } else if let systemImage {
                    Image(systemName: systemImage)
                }
                Text(title).font(.headline)
            }
            .frame(maxWidth: .infinity)
            .frame(minHeight: 58)
            .background(SBTColor.primary, in: RoundedRectangle(cornerRadius: SBTMetrics.cornerRadius))
            .foregroundStyle(.white)
        }
        .buttonStyle(.plain)
        .disabled(isBusy)
    }
}

/// The secondary action button, outlined rather than filled.
struct SecondaryButton: View {

    let title: String
    var systemImage: String?
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            HStack(spacing: 10) {
                if let systemImage { Image(systemName: systemImage) }
                Text(title).font(.body.weight(.semibold))
            }
            .frame(maxWidth: .infinity)
            .frame(minHeight: SBTMetrics.touchTarget)
            .background(SBTColor.surface, in: RoundedRectangle(cornerRadius: SBTMetrics.cornerRadius))
            .overlay(
                RoundedRectangle(cornerRadius: SBTMetrics.cornerRadius)
                    .stroke(SBTColor.outline, lineWidth: 1.5)
            )
            .foregroundStyle(SBTColor.ink)
        }
        .buttonStyle(.plain)
    }
}
